### ISSUE TITLE
ray.tune - Lifecycle of a trial - where to set max_retries?

### ISSUE THREAD

**Issue by viotemp1** [ray-project/ray#11782]
Hello,

Where can I set max_retries when using tune.run? I'm running tune.run with tensorflow on a single workstation with two GPUs.

I'm trying to use this feature from trial lifecycle:
ERRORED: If a running trial throws an exception, Tune will catch that exception and mark the trial as errored. Note that exceptions can be propagated from an actor to the main Tune driver process. If **max_retries** is set, Tune will set the trial back into “PENDING” and later start it from the last checkpoint.

This feature it's important for me to retry later failed trials that might not have enough GPU memory to run in parallel

Regards,
Viorel

**Comment by krfricke** [comment]
Hi @viotemp1, I think the parameter was renamed to `max_failures` - could you try setting this in `tune.run()`? E.g.

```
tune.run(
    trainable,
    # ...
    max_failures=4)
```

If you set it to `-1` it will retry an infinite number of times.

**Comment by viotemp1** [comment]
Hi @krfricke,

I use max_failures=4 but I do not see the tasks going to PENDING. They are going to ERROR state after 4 retries.
I was looking for this behaviour:  If max_retries is set, Tune will set the trial back into “PENDING” and later start it from the last checkpoint.

So the tasks which require most GPU memory (end eventually failed because of this) will be left as PENDING and eventually retried later.

As a workaround, I will try to use for now resume="ERRORED_ONLY" for these failing tasks and decrease max_concurrent to see how it goes.
Regards
Viorel

**Comment by viotemp1** [comment]
Now I see there is Discussion section for this type of questions.
Should I close here on open in Discussion section?
Regards,

**Comment by krfricke** [comment]
It's fine to leave it here for the time being. Could you provide us with a bit more context, e.g. your `tune.run()` command? Specifically it would be helpful to see if you're requesting resources correctly. Do you pass `resources_per_trial` to `tune.run()`?

If not, you should try it:

```
tune.run(
    trainable,
    # ...
    resources_per_trial={"gpu": 1},
    num_failures=4)
```

This will request 1 GPU per trial, and it will only schedule trials if a GPU is free and available. You can pass fractional GPUs, too (like `gpu: 0.5`). This should only schedule trials once the GPU is free.

**Comment by viotemp1** [comment]
Hello,

I'm doing resource allocations (general and per trial) - see bellow.
The ideea is that for most of trials I do not have errors, but some combinations of higher no of filters/units for conv2d/dense/LSTM from my model - give OOM when 2 tasks are running in parallel on one GPU.
Failed/Errored trials I would run later on with resume and only one trial not more - this is the solution I'm trying now.

I was thinking that ray tune can mark these failing as PENDING and try to run it at the end again. Maybe I was wrong.

Partially source code:

Ray init:
        ray.init(
            log_to_driver=False,
            configure_logging=ray_debug, # False
            include_dashboard=False,
            num_cpus=multiprocessing.cpu_count(), # 12
            num_gpus=2 if (len(os.environ["CUDA_VISIBLE_DEVICES"])) >= 3 else 1, # CUDA_VISIBLE_DEVICES = 1 now
        )

ConcurrencyLimiter(search_alg, max_concurrent=max_concurrent) # 2

Ray tune:
        analysis = tune.run(
            tune.with_parameters(
                tune_fn,
                data_dir=data_dir,
            ),
            config=hyperparameter_space,
            name=experiment_name,
            search_alg=search_alg, # HyperOptSearch
            stop=stop,
            scheduler=scheduler, # MedianStoppingRule
            callbacks=[ray_Callback()],
            **experiment_config_metrics,
        )

Resources per trial specified in experiment_config_metrics:
{'verbose': 1,
 'resume': False,
 'metric': 'val_my_Metric',
 'mode': 'min',
 'num_samples': 200,
 'checkpoint_freq': 0,
 'keep_checkpoints_num': None,
 'checkpoint_score_attr': None,
 'checkpoint_at_end': True,
 'max_failures': 2,
 'raise_on_failed_trial': False,
 'queue_trials': True,
 'local_dir': '/home/.../PythonProjects/test_ray/logs1/',
 'resources_per_trial': {'cpu': 5, 'gpu': 0.45}
}

**Comment by viotemp1** [comment]
Just to make the question a little bit more clear: I set max_concurrency to 2, one task use 0.45 GPU and the next one has 0.55 GPU available, but does not fit into this (because the model is larger). The last task is not going to PENDING, but to ERROR after few retries.
![Screenshot 2020-11-03 at 16 23 15](https://user-images.githubusercontent.com/34300519/97997367-8fb10900-1df1-11eb-8163-26a6ae6d39cc.png)
![Screenshot 2020-11-03 at 16 24 15](https://user-images.githubusercontent.com/34300519/97997389-950e5380-1df1-11eb-851b-e0c80047faac.png)

**Comment by richardliaw** [comment]
Can you please post the error logs? The trial may be in "PENDING" for a very short period of time if there is space on the Ray cluster.

**Comment by viotemp1** [comment]
Hello,

You might me right, maybe I do not see it in Jupiter cell output  - refresh is not very frequent. I’m using verbose=1
Are you talking about experiment_state….json file? There are some lines there with "status": “PENDING” indeed.

Or I should look for another log file? Sorry, not sure about log filenames and their locations., that’s why I’m asking. 
Should I set verbose to 2?

Regards,
Viorel

**Comment by viotemp1** [comment]
I was looking for longer time and I see tasks in PENDING.
We can close this question.
Regards,
Viorel

**Comment by krfricke** [comment]
Great, let us know if there's anything else!

**Comment by viotemp1** [comment]
I did open some other threads in Discussion pane, thanks. 
Anyhow, I start to like ray(tune) more and more. Very good job!
